### PR TITLE
store: add lock for runtime stats to fix panic caused by concurrent execution

### LINKED ISSUE #18985: panic in mergeRegionRequestStats
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Reproduce:

```sql
create table t1 (id int, b int,  unique index (id));
create table t2 (id int, b int,  unique index (id));
insert into t1 values  (1,1);
insert into t2 values  (1,1);
```

then run:

```sql
select t1.*,t2.* from t1 left join t2 on t1.id=t2.id where t1.id=1;  -- This maybe panic occasionally.
```

### 3. What did you see instead (Required)

In release-4.0 branch, hash: `df16c56c51fbc6710431ff1ee7bbac23c610e924`, the panic stack is below:

```go
[2020/08/04 15:41:52.688 +00:00] [ERROR] [misc.go:91] ["panic in the recoverable goroutine"] [r="\"assignment to entry in nil map\""] ["stack trace"="github.com/pingcap/tidb/util.WithRecovery.func1
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/util/misc.go:93
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:679
runtime.mapassign
	/usr/local/go/src/runtime/map.go:576
github.com/pingcap/tidb/store/tikv.(*tikvSnapshot).mergeRegionRequestStats
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:597
github.com/pingcap/tidb/store/tikv.(*tikvSnapshot).get.func2
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:369
github.com/pingcap/tidb/store/tikv.(*tikvSnapshot).get
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:424
github.com/pingcap/tidb/store/tikv.(*tikvSnapshot).Get
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/store/tikv/snapshot.go:326
github.com/pingcap/tidb/executor.(*PointGetExecutor).get
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/executor/point_get.go:319
github.com/pingcap/tidb/executor.(*PointGetExecutor).getAndLock
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/executor/point_get.go:264
github.com/pingcap/tidb/executor.(*PointGetExecutor).Next
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/executor/point_get.go:219
github.com/pingcap/tidb/executor.Next
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/executor/executor.go:250
github.com/pingcap/tidb/executor.(*HashJoinExec).fetchProbeSideChunks
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/executor/join.go:203
github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndProbeHashTable.func1
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/executor/join.go:316
github.com/pingcap/tidb/util.WithRecovery
	/home/jenkins/agent/workspace/tidb_release-4.0/go/src/github.com/pingcap/tidb/util/misc.go:96"]
```

### 4. Affected version (Required)

Master & latest release 4.0

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

fix #18985

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit Test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

-  Add lock for runtime stats to fix panic caused by concurrent execution.

### CODE REVIEW

File: store/tikv/snapshot.go [hunk 1]
```diff
@@ -462,7 +464,9 @@ func (s *tikvSnapshot) DelOption(opt kv.Option) {
 	case kv.ReplicaRead:
 		s.replicaRead = kv.ReplicaReadLeader
 	case kv.CollectRuntimeStats:
+		s.mu.Lock()
 		s.stats = nil
+		s.mu.Unlock()
```
Comment (qw4990) [inline store/tikv/snapshot.go:469]: Almost LGTM for me, but please put the field `stats` into `mu` in the definition of `tikvSnapshot`.

Comment (crazycs520) [reply store/tikv/snapshot.go:469]: great, done

**Pull request merged**